File: clients/dotnet/Amazon.DynamoDBStreams.Consumer/Worker.cs
using System.Diagnostics;
using System.Text.Json;

namespace Amazon.DynamoDBStreams.Consumer;

/// <summary>
/// Where a freshly-seeded shard (no checkpoint) begins reading:
/// <see cref="InitialPosition.TrimHorizon"/> (default) or <see cref="InitialPosition.Latest"/>.
/// </summary>
public enum InitialPosition
{
    /// <summary>Read the shard from the oldest available record (default).</summary>
    TrimHorizon,

    /// <summary>Read only records written after the consumer starts.</summary>
    Latest,
}

/// <summary>Configuration for a <see cref="Worker"/>.</summary>
public sealed class WorkerConfig
{
    /// <summary>The DynamoDB Streams ARN to consume (required).</summary>
    public string StreamArn { get; set; } = "";

    /// <summary>The DynamoDB table storing shard leases + checkpoints (required).</summary>
    public string LeaseTable { get; set; } = "";

    /// <summary>The record processor (required).</summary>
    public IRecordProcessor Processor { get; set; } = null!;

    /// <summary>Unique worker identity for lease ownership. Optional.</summary>
    public string? Owner { get; set; }

    /// <summary>AWS region. Optional (falls back to the standard AWS environment).</summary>
    public string? Region { get; set; }

    /// <summary>How attribute values are surfaced. Defaults to <see cref="RecordFormat.Native"/>.</summary>
    public RecordFormat RecordFormat { get; set; } = RecordFormat.Native;

    /// <summary>Maximum shard leases held at once. Optional.</summary>
    public int? MaxLeases { get; set; }

    /// <summary>Cap on shards processed concurrently (opt-in). Unset = one slot per shard.
    /// Bounds concurrent delivery so footprint stays O(max) as shard count grows;
    /// preserves at-least-once + per-item + per-shard ordering.</summary>
    public int? MaxProcessingConcurrency { get; set; }

    /// <summary>Lease duration in milliseconds. Optional.</summary>
    public long? LeaseDurationMs { get; set; }

    /// <summary>Idle poll backoff in milliseconds. Optional.</summary>
    public long? PollIntervalMs { get; set; }

    /// <summary>Coordination cycle interval in milliseconds. Optional.</summary>
    public long? CycleIntervalMs { get; set; }

    /// <summary>
    /// Where to start reading a shard that has no checkpoint. Optional.
    /// Allowed values: <c>TRIM_HORIZON</c> (default) reads from the oldest
    /// available record; <c>LATEST</c> reads only records written after the
    /// worker starts. Case-insensitive.
    /// </summary>
    public InitialPosition? InitialPosition { get; set; }

    /// <summary>Explicit sidecar binary path (overrides discovery). Optional.</summary>
    public string? SidecarPath { get; set; }

    /// <summary>Full launch argv (tests / custom launch; overrides discovery). Optional.</summary>
    public IReadOnlyList<string>? SidecarCmd { get; set; }
}

/// <summary>
/// A JVM-free DynamoDB Streams consumer. Spawns the shared Rust sidecar and
/// delivers ordered, checkpointed change records to an <see cref="IRecordProcessor"/>
/// over the JSON-Lines wire protocol.
/// </summary>
public sealed class Worker
{
    private readonly WorkerConfig _config;
    private Process? _proc;
    private volatile bool _closed;

    /// <summary>Create a worker from the given configuration.</summary>
    public Worker(WorkerConfig config)
    {
        if (config is null || string.IsNullOrEmpty(config.StreamArn)
            || string.IsNullOrEmpty(config.LeaseTable) || config.Processor is null)
        {
            throw new ArgumentException("StreamArn, LeaseTable and Processor are required.", nameof(config));
        }
        _config = config;
    }

    /// <summary>
    /// Run until the sidecar shuts down (stream fully consumed, <see cref="Stop"/>
    /// called, or fatal error). Returns the sidecar's exit code.
    /// </summary>
    public async Task<int> RunAsync()
    {
        var argv = _config.SidecarCmd is { Count: > 0 } cmd
            ? new List<string>(cmd)
            : new List<string> { await Sidecar.DiscoverAsync(_config.SidecarPath).ConfigureAwait(false) };

        var psi = new ProcessStartInfo
        {
            FileName = argv[0],
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = false, // inherit → sidecar logs to our stderr
            UseShellExecute = false,
        };
        for (var i = 1; i < argv.Count; i++)
        {
            psi.ArgumentList.Add(argv[i]);
        }
        ApplyEnv(psi);

        _proc = new Process { StartInfo = psi };
        _proc.Start();
        var stdin = _proc.StandardInput;
        var stdout = _proc.StandardOutput;

        Send(stdin, new { type = "ready" });

        string? line;
        while ((line = await stdout.ReadLineAsync().ConfigureAwait(false)) != null)
        {
            line = line.Trim();
            if (line.Length == 0)
            {
                continue;
            }

            JsonDocument doc;
            try
            {
                doc = JsonDocument.Parse(line);
            }
            catch (JsonException)
            {
                continue; // ignore malformed / non-protocol noise
            }

            using (doc)
            {
                var root = doc.RootElement;
                if (root.ValueKind != JsonValueKind.Object
                    || !root.TryGetProperty("type", out var typeEl))
                {
                    continue;
                }
                switch (typeEl.GetString())
                {
                    case "records":
                        HandleRecords(root, stdin);
                        break;
                    case "shard_complete":
                        if (root.TryGetProperty("shard", out var sc))
                        {
                            _config.Processor.ShardEnded(sc.GetString() ?? "");
                        }
                        break;
                    case "lease_lost":
                        if (root.TryGetProperty("shard", out var ll))
                        {
                            _config.Processor.LeaseLost(ll.GetString() ?? "");
                        }
                        break;
                    case "shutdown_requested":
                        if (root.TryGetProperty("shard", out var sr))
                        {
                            _config.Processor.ShutdownRequested(sr.GetString() ?? "");
                        }
                        break;
                    case "shutdown":
                        StopInternal(stdin);
                        break;
                    default:
                        break;
                }
            }
        }

        _proc.WaitForExit();
        _closed = true;
        return _proc.ExitCode;
    }

    /// <summary>Request a graceful shutdown; <see cref="RunAsync"/> resolves once the sidecar exits.</summary>
    public void Stop()
    {
        if (_proc is { } p && !_closed)
        {
            try
            {
                Send(p.StandardInput, new { type = "stop" });
            }
            catch (IOException)
            {
                // pipe already gone
            }
        }
    }

    private void HandleRecords(JsonElement root, StreamWriter stdin)
    {
        var shard = root.TryGetProperty("shard", out var sh) ? sh.GetString() ?? "" : "";
        var lastSeq = root.TryGetProperty("last_seq", out var ls) ? ls.GetString() : null;

        var records = new List<Record>();
        if (root.TryGetProperty("records", out var recs) && recs.ValueKind == JsonValueKind.Array)
        {
            foreach (var r in recs.EnumerateArray())
            {
                records.Add(RecordFromWire(shard, r, _config.RecordFormat));
            }
        }

        _config.Processor.ProcessRecords(records);
        Send(stdin, new { type = "checkpoint", shard, seq = lastSeq });
    }

    private static Record RecordFromWire(string shard, JsonElement w, RecordFormat fmt)
    {
        string? Str(string prop) =>
            w.TryGetProperty(prop, out var el) && el.ValueKind == JsonValueKind.String ? el.GetString() : null;

        Dictionary<string, object?>? Image(string prop)
        {
            if (!w.TryGetProperty(prop, out var el) || el.ValueKind != JsonValueKind.Object)
            {
                return null;
            }
            return fmt == RecordFormat.Sdk
                ? SdkAttributeValues.DecodeItem(el)
                : AttributeValueConverter.DecodeItem(el, fmt);
        }

        return new Record
        {
            ShardId = shard,
            EventName = Str("event_name"),
            SequenceNumber = Str("sequence_number"),
            StreamViewType = Str("stream_view_type"),
            Keys = Image("keys") ?? new Dictionary<string, object?>(),
            NewImage = Image("new_image"),
            OldImage = Image("old_image"),
        };
    }

    private void ApplyEnv(ProcessStartInfo psi)
    {
        var e = psi.Environment;
        e["DDB_STREAMS_CONSUMER_STREAM_ARN"] = _config.StreamArn;
        e["DDB_STREAMS_CONSUMER_LEASE_TABLE"] = _config.LeaseTable;
        if (!string.IsNullOrEmpty(_config.Owner))
        {
            e["DDB_STREAMS_CONSUMER_OWNER"] = _config.Owner!;
        }
        if (!string.IsNullOrEmpty(_config.Region))
        {
            e["AWS_REGION"] = _config.Region!;
        }
        if (_config.MaxLeases is { } ml)
        {
            e["DDB_STREAMS_CONSUMER_MAX_LEASES"] = ml.ToString();
        }
        if (_config.MaxProcessingConcurrency is { } mpc)
        {
            e["DDB_STREAMS_CONSUMER_MAX_PROCESSING_CONCURRENCY"] = mpc.ToString();
        }
        if (_config.LeaseDurationMs is { } ld)
        {
            e["DDB_STREAMS_CONSUMER_LEASE_DURATION_MS"] = ld.ToString();
        }
        if (_config.PollIntervalMs is { } pi)
        {
            e["DDB_STREAMS_CONSUMER_POLL_INTERVAL_MS"] = pi.ToString();
        }
        if (_config.CycleIntervalMs is { } ci)
        {
            e["DDB_STREAMS_CONSUMER_CYCLE_INTERVAL_MS"] = ci.ToString();
        }
        if (_config.InitialPosition is { } ip)
        {
            e["DDB_STREAMS_CONSUMER_INITIAL_POSITION"] = ip switch
            {
                InitialPosition.Latest => "LATEST",
                _ => "TRIM_HORIZON",
            };
        }
    }

    private void StopInternal(StreamWriter stdin)
    {
        if (_closed)
        {
            return;
        }
        try
        {
            Send(stdin, new { type = "stop" });
        }
        catch (IOException)
        {
            // pipe gone
        }
        try
        {
            stdin.Close();
        }
        catch (IOException)
        {
            // already closed
        }
    }

    private static void Send(StreamWriter stdin, object msg)
    {
        // Best-effort: the sidecar may have already emitted its final message and
        // exited (closing its stdin), so a write can race into a broken pipe. That
        // is a normal shutdown, not an error — the read loop observes stdout EOF
        // and stops. Mirrors the other clients (Python swallows BrokenPipeError).
        try
        {
            stdin.Write(JsonSerializer.Serialize(msg));
            stdin.Write('\n');
            stdin.Flush();
        }
        catch (IOException)
        {
            // sidecar stdin already closed
        }
        catch (ObjectDisposedException)
        {
            // stdin already disposed
        }
    }
}

File: clients/java/src/main/java/com/amazon/dynamodbstreams/consumer/Worker.java
package com.amazon.dynamodbstreams.consumer;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * A JVM-free-core DynamoDB Streams consumer. Spawns the shared Rust sidecar and
 * delivers ordered, checkpointed change records to a {@link RecordProcessor}
 * over the JSON-Lines wire protocol. The coordination logic (shard discovery,
 * leasing, ordering, checkpointing) runs in the native sidecar, not the JVM.
 */
public final class Worker {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final WorkerConfig config;
    private volatile Process proc;
    private volatile boolean closed;

    public Worker(WorkerConfig config) {
        this.config = config;
    }

    /**
     * Run until the sidecar shuts down (stream fully consumed, {@link #stop()}
     * called, or fatal error). Returns the sidecar's exit code.
     */
    public int run() throws IOException, InterruptedException {
        List<String> argv = (config.sidecarCmd != null && !config.sidecarCmd.isEmpty())
                ? new ArrayList<>(config.sidecarCmd)
                : List.of(Sidecar.discover(config.sidecarPath));

        ProcessBuilder pb = new ProcessBuilder(argv);
        pb.redirectError(ProcessBuilder.Redirect.INHERIT); // sidecar logs to our stderr
        applyEnv(pb);

        proc = pb.start();
        OutputStream stdin = proc.getOutputStream();
        BufferedReader stdout = new BufferedReader(
                new InputStreamReader(proc.getInputStream(), StandardCharsets.UTF_8));

        send(stdin, readyMessage());

        String line;
        while ((line = stdout.readLine()) != null) {
            line = line.trim();
            if (line.isEmpty()) {
                continue;
            }
            JsonNode msg;
            try {
                msg = MAPPER.readTree(line);
            } catch (IOException e) {
                continue; // ignore malformed / non-protocol noise
            }
            if (!msg.isObject() || !msg.hasNonNull("type")) {
                continue;
            }
            String type = msg.get("type").asText();
            switch (type) {
                case "records":
                    handleRecords(msg, stdin);
                    break;
                case "shard_complete":
                    config.processor.shardEnded(msg.path("shard").asText(""));
                    break;
                case "lease_lost":
                    config.processor.leaseLost(msg.path("shard").asText(""));
                    break;
                case "shutdown_requested":
                    config.processor.shutdownRequested(msg.path("shard").asText(""));
                    break;
                case "shutdown":
                    stopInternal(stdin);
                    break;
                default:
                    break;
            }
        }

        int code = proc.waitFor();
        closed = true;
        return code;
    }

    /** Request a graceful shutdown; {@link #run()} returns once the sidecar exits. */
    public void stop() {
        Process p = proc;
        if (p != null && !closed) {
            try {
                send(p.getOutputStream(), stopMessage());
            } catch (IOException ignored) {
                // pipe already gone
            }
        }
    }

    private void handleRecords(JsonNode root, OutputStream stdin) throws IOException {
        String shard = root.path("shard").asText("");
        String lastSeq = root.path("last_seq").asText(null);

        List<Record> records = new ArrayList<>();
        JsonNode recs = root.get("records");
        if (recs != null && recs.isArray()) {
            for (JsonNode r : recs) {
                records.add(recordFromWire(shard, r, config.recordFormat));
            }
        }

        config.processor.processRecords(records);

        ObjectNode ck = MAPPER.createObjectNode();
        ck.put("type", "checkpoint");
        ck.put("shard", shard);
        ck.put("seq", lastSeq);
        send(stdin, ck);
    }

    private static Record recordFromWire(String shard, JsonNode w, RecordFormat fmt) {
        return new Record(
                shard,
                text(w, "event_name"),
                text(w, "sequence_number"),
                text(w, "stream_view_type"),
                image(w, "keys", fmt),
                image(w, "new_image", fmt),
                image(w, "old_image", fmt));
    }

    private static String text(JsonNode w, String field) {
        JsonNode n = w.get(field);
        return (n != null && n.isTextual()) ? n.asText() : null;
    }

    private static Map<String, Object> image(JsonNode w, String field, RecordFormat fmt) {
        JsonNode n = w.get(field);
        if (n == null || !n.isObject()) {
            return null;
        }
        return fmt == RecordFormat.SDK
                ? SdkAttributeValues.decodeItem(n)
                : AttributeValues.decodeItem(n, fmt);
    }

    private void applyEnv(ProcessBuilder pb) {
        Map<String, String> e = pb.environment();
        e.put("DDB_STREAMS_CONSUMER_STREAM_ARN", config.streamArn);
        e.put("DDB_STREAMS_CONSUMER_LEASE_TABLE", config.leaseTable);
        if (config.owner != null && !config.owner.isEmpty()) {
            e.put("DDB_STREAMS_CONSUMER_OWNER", config.owner);
        }
        if (config.region != null && !config.region.isEmpty()) {
            e.put("AWS_REGION", config.region);
        }
        if (config.maxLeases != null) {
            e.put("DDB_STREAMS_CONSUMER_MAX_LEASES", config.maxLeases.toString());
        }
        if (config.maxProcessingConcurrency != null) {
            e.put("DDB_STREAMS_CONSUMER_MAX_PROCESSING_CONCURRENCY", config.maxProcessingConcurrency.toString());
        }
        if (config.leaseDurationMs != null) {
            e.put("DDB_STREAMS_CONSUMER_LEASE_DURATION_MS", config.leaseDurationMs.toString());
        }
        if (config.pollIntervalMs != null) {
            e.put("DDB_STREAMS_CONSUMER_POLL_INTERVAL_MS", config.pollIntervalMs.toString());
        }
        if (config.cycleIntervalMs != null) {
            e.put("DDB_STREAMS_CONSUMER_CYCLE_INTERVAL_MS", config.cycleIntervalMs.toString());
        }
        if (config.initialPosition != null) {
            e.put("DDB_STREAMS_CONSUMER_INITIAL_POSITION", config.initialPosition.name());
        }
    }

    private void stopInternal(OutputStream stdin) {
        if (closed) {
            return;
        }
        try {
            send(stdin, stopMessage());
        } catch (IOException ignored) {
            // pipe gone
        }
        try {
            stdin.close();
        } catch (IOException ignored) {
            // already closed
        }
    }

    private static ObjectNode readyMessage() {
        ObjectNode n = MAPPER.createObjectNode();
        n.put("type", "ready");
        return n;
    }

    private static ObjectNode stopMessage() {
        ObjectNode n = MAPPER.createObjectNode();
        n.put("type", "stop");
        return n;
    }

    private static void send(OutputStream stdin, JsonNode msg) throws IOException {
        byte[] bytes = (MAPPER.writeValueAsString(msg) + "\n").getBytes(StandardCharsets.UTF_8);
        stdin.write(bytes);
        stdin.flush();
    }
}
